### DISCUSSION TITLE
Module core should recognize the reset action not only install/uninstall

### DISCUSSION THREAD

**Discussion by ShaiMagal** [PrestaShop/PrestaShop#37748] [Ideas]
### Prerequisites

- [x] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [x] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

Module.php now have function install() and uninstall() and it's working well.

But there is a specific scenario with **reset** module. If you reset module, it simply does uninstall and install. 
But we need to know IF the admin clicked reset or not.

**Real use case:**
- We have module with some sql table.
- This table contains very important data.
- If customer **uninstalls** module, we delete the table, because we "know" that the customer probably will stop using the module.
- But if customer **resets** module, we don't want to delete the sql table, becase we know that it's reinstallation and customer will continue with usage of this data.

Now in current state, we aren't able to get this information.

**reinstallation** = the customer says that he will continue to use the module, so we do not delete the table
**uninstallation** = the customer says that he will no longer use the module, so we delete the table

**We need to condition it somehow in the install/uninstall function to recognize that it is a reset and not an uninstall.**

### Expected behavior

Click on "reset" should be "flagged" for install() and uninstall() job for module developer to handle this.

### Steps to reproduce

1. go to section module
2. reset any module
3. now module don't know if it were reset, it just reinstall whole module, but don't tell module.

### PrestaShop version(s) where the bug happened

9.0.0

### PHP version(s) where the bug happened

8.3

### Your company or customer's name goes here (if applicable).

https://www.openservis.cz/

**Comment by matks** [comment]
Hi @ShaiMagal I am not sure to understand

- in the backoffice page I think in the URL you can see the action used, "reset" in this case
- in the class ModuleManager ; there is a function `reset()`

So when you say `Now in current state, we aren't able to get this information.` I don't understand. I can clearly see what is a reset and what is not.

https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Module/ModuleManager.php#L282

**Comment by ShaiMagal** [comment]
@matks When I click on "reset" in BO, how module can get know, it's really reseted?
Use case is, if I click "reset" it do "uninstall" and then "install".

I need to know in function "install" there were any reset. How this can be done?

I see function 
**public function reset(string $name, bool $keepData = false): bool**
"$keepData" looks good, but, how we can do this?

Do you have any module example what is using this? I never saw any module with reset feature.

@Hlavtox can you please explain to matks what I am solving? :-D I think, there is language barrier :-D

**Comment by kpodemski** [comment]
It’s the issue we’ve discussed with @jolelievre and @Progi1984 recently, the reset action is indeed not working correctly

@matks module manager doesn’t know about the choice for the $reset parameter, so eventually, it is useless

**Comment by jolelievre** [comment]
@ShaiMagal this may have been fixed by this PR https://github.com/PrestaShop/PrestaShop/pull/37463/files#diff-916b3331799e0e2b74d59aba5995f02bebd746d989b950cff499cb431b2e8e57R301

Although it was not related, via some errors triggered by a more recent PHPUnit version I noticed that the `reset` method existence was not checked on the appropriate object

**Comment by paulnoelcholot** [comment]
Hello @ShaiMagal,

If it's ok for you can I close the issue?

Thanks for your feedback!

**Comment by ShaiMagal** [comment]
@jolelievre @paulnoelcholot how this PR will help me (this issue) with detection of "reset"?
https://github.com/PrestaShop/PrestaShop/pull/37463/files#diff-916b3331799e0e2b74d59aba5995f02bebd746d989b950cff499cb431b2e8e57R301

@kpodemski @matks @Hlavtox 
"module manager doesn’t know about the choice for the $reset parameter, so eventually, it is useless"
useless? realy?
The fact that no one has done it yet doesn't mean it's useless. Do you understand the reason why I want it? (see my issue) Or should I explain more/better?

**Comment by paulnoelcholot** [comment]
I think this is more of a feature. I'm moving the issue to discussion.

**Comment by Hlavtox** [comment]
@ShaiMagal What Krystian meant is that the parameter is not implemented, by a mistake probably. That's why it's currently useless.

Ping @jolelievre @matks @kpodemski - it's pretty important topic, because modules need to know if it's a clean install or a reset. If they should wipe all data or not. And I think that Dominik knows his shit, with over 200 modules on the market. :-)))
